src/webots/vrml/WbProtoManager.cpp

```diff
@@ -105,85 +107,116 @@ void WbProtoManager::readModel(WbTokenizer *tokenizer, const QString &worldPath)
+        return NULL;
+      }
+    } else
+      modelPath = WbUrl::computePath(protoDeclaration);
+  } else
+    modelPath = WbUrl::computePath(protoDeclaration);
+
+  // determine prefix and disk location from modelPath
+  const QString modelDiskPath = WbUrl::isWeb(modelPath) ? WbNetwork::instance()->get(modelPath) : modelPath;
+  const QString prefix = WbUrl::computePrefix(modelPath);  // used to retrieve remote assets (replaces webots:// in the body)
 
-    WbProtoModel *model = readModel(QFileInfo(url).absoluteFilePath(), worldPath, url, baseTypeList);
+  if (QFileInfo(modelDiskPath).exists() && !modelPath.isEmpty()) {
+    WbProtoModel *model = readModel(modelPath, worldPath, prefix, baseTypeList);
     if (model == NULL)  // can occur if the PROTO contains errors
       return NULL;
     mModels << model;
     model->ref();
     return model;
-  } else {  // check if the PROTO is locally available, if so notify the user that an EXTERNPROTO declaration is needed
-    // check in the project's protos directory
-    QDirIterator it(WbProject::current()->protosPath(), QStringList() << "*.proto", QDir::Files, QDirIterator::Subdirectories);
-    while (it.hasNext()) {
-      const QString &protoPath = it.next();
-      if (modelName == QFileInfo(protoPath).baseName()) {
-        const QString errorMessage =
-          tr("PROTO '%1' is available locally but was not declared, please do so by adding the following line to "
-             "the world file: EXTERNPROTO \"../protos/%2\"")
-            .arg(modelName)
-            .arg(QFileInfo(protoPath).fileName());
-
-        if (!mUniqueErrorMessages.contains(errorMessage)) {
-          WbLog::error(errorMessage);
-          mUniqueErrorMessages << errorMessage;
-        }
-        return NULL;
-      }
-    }
-    // check in the extra project directories
-    foreach (const WbProject *project, *WbProject::extraProjects()) {
-      QDirIterator it(project->protosPath(), QStringList() << "*.proto", QDir::Files, QDirIterator::Subdirectories);
-      while (it.hasNext()) {
-        const QString &protoPath = it.next();
-        if (modelName == QFileInfo(protoPath).baseName()) {
-          const QString errorMessage =
-            tr("PROTO '%1' is available locally but was not declared, please do so by adding the following line to "
-               "the world file: EXTERNPROTO \"%2\"")
-              .arg(modelName)
-              .arg(protoPath);
-
-          if (!mUniqueErrorMessages.contains(errorMessage)) {
-            WbLog::error(errorMessage);
-            mUniqueErrorMessages << errorMessage;
-          }
-
-          return NULL;
-        }
-      }
-    }
   }
 
   return NULL;
 }
 
+QString WbProtoManager::findExternProtoDeclarationInFile(const QString &url, const QString &modelName) {
+  if (url.isEmpty())
+    return QString();
+
+  QString path = url;
+  if (WbUrl::isWeb(path))
+    path = WbNetwork::instance()->get(path);
```

Comment:
This could be also done in one line:
```suggestion
  const QString path = WbUrl::isWeb(path) ? WbNetwork::instance()->get(url) : url;
```